### ISSUE TITLE
Tests & SSLv3

### ISSUE THREAD

**Issue by MSP-Greg** [eventmachine/eventmachine#786]
After building with OpenSSL 1.1.0, I ran the tests, and it seems that several are using SSLv3. I suspect that in the future, more and more OpenSSL packages/libraries will disable SSLv3.

Checking my system, I showed:
```
irb(main):003:0> OpenSSL::SSL::SSLContext::METHODS
=> [:TLSv1, :TLSv1_server, :TLSv1_client, :TLSv1_1, :TLSv1_1_server, :TLSv1_1_client,
    :TLSv1_2, :TLSv1_2_server, :TLSv1_2_client, :SSLv23, :SSLv23_server, :SSLv23_client]
```

I'd be happy to look at updating the tests, but I'm not sure whether to 'skip' if SSLv3 is unavailable, 
change to another protocol, or remove SSLv3 if the test is checking more than one...

**Comment by sodabrew** [comment]
Any tests that are explicitly using SSLv3 can be switched to an appropriate TLS version.

Your system shows that SSLv23 (which means "allow SSL version 2 or version 3") is available, at the constants are.

**Comment by MSP-Greg** [comment]
@sodabrew Thanks for the reply.  I have little knowledge of OpenSSL, but:

```
openssl s_client -connect google.com:443 -ssl3
=> s_client: Option unknown option -ssl3
=> s_client: Use -help for summary.
```

My understanding is that I should get a different message if my system has SSLv3 enabled but the server does not.

Additionally, both the following give good responses
```
openssl s_client -connect google.com:443
openssl s_client -connect google.com:443 -tls1_2
```

**Comment by MSP-Greg** [comment]
@sodabrew 

I rebuild my MinGW OpenSSL package with the recent 1.1.0f release.  I built with `--api=1.0.0`, which I believe removes any 0.9x API's that are not current.  the following is output by the build --

```
Configuring OpenSSL version 1.1.0f (0x1010006fL)
    no-asan         [default]  OPENSSL_NO_ASAN
    no-crypto-mdebug [default]  OPENSSL_NO_CRYPTO_MDEBUG
    no-crypto-mdebug-backtrace [default]  OPENSSL_NO_CRYPTO_MDEBUG_BACKTRACE
    no-ec_nistp_64_gcc_128 [default]  OPENSSL_NO_EC_NISTP_64_GCC_128
    no-egd          [default]  OPENSSL_NO_EGD
    no-fuzz-afl     [default]  OPENSSL_NO_FUZZ_AFL
    no-fuzz-libfuzzer [default]  OPENSSL_NO_FUZZ_LIBFUZZER
    no-heartbeats   [default]  OPENSSL_NO_HEARTBEATS
    no-md2          [default]  OPENSSL_NO_MD2 (skip dir)
    no-msan         [default]  OPENSSL_NO_MSAN
    no-rc5          [default]  OPENSSL_NO_RC5 (skip dir)
    no-sctp         [default]  OPENSSL_NO_SCTP
    no-ssl-trace    [default]  OPENSSL_NO_SSL_TRACE
    no-ssl3         [default]  OPENSSL_NO_SSL3
    no-ssl3-method  [default]  OPENSSL_NO_SSL3_METHOD
    no-ubsan        [default]  OPENSSL_NO_UBSAN
    no-unit-test    [default]  OPENSSL_NO_UNIT_TEST
    no-weak-ssl-ciphers [default]  OPENSSL_NO_WEAK_SSL_CIPHERS
```

> Any tests that are explicitly using SSLv3 can be switched to an appropriate TLS version.

I don't normally use EM with SSL, but if I get back to testing, would you accept PR's that remove the hard-coded SSLv3 tests?  Several days ago I switched several to SSLv23, and the affected tests passed.

FYI, both the MinGW package and ruby trunk pass all OpenSSL tests...

**Comment by sodabrew** [comment]
Yes, thank you - migrating the tests would be welcomed as a PR.

**Comment by MSP-Greg** [comment]
@sodabrew 

Well, went for two birds --

1. Updated SSL tests
2. Ruby 2.4.1 build & test are running, with SSL, on [Appveyor](https://ci.appveyor.com/project/MSP-Greg/eventmachine)

Two failures (L62, L75), both involving a return error message, where the assert is using a regex match.

The strings returned are:
```
AppVeyor
"unable to resolve address: No such host is known. "
"invalid bind address: No such host is known. "

My system
"unable to resolve address: The requested name is valid, but no data of the requested type was found. "
"invalid bind address: The requested name is valid, but no data of the requested type was found. "
```

If those are correct returns, I can update the regex. Or, given the differences, maybe just test from the start to the colon?

Re SSL tests, I added some TLS tests, added omits depending on whether SSL and/or TLS was enabled, etc.

I had to pend one test that generated a segfault (L132).

The commits aren't pretty right now, spread across two branches.  If this look okays for a PR, one big commit, or split?  Maybe one PR for SSL, and one for Win tests and Appveyor?  Whatever you'd prefer...

FYI, since I work mostly with trunk, and very few people have updated bundler to work with it, it gets in my way more than it helps. Hence, no bundler used in Appveyor.

**Comment by MSP-Greg** [comment]
Updated the tests, using bundler.  Pass on [Appveyor](https://ci.appveyor.com/project/MSP-Greg/eventmachine)

**Comment by sodabrew** [comment]
Great! Are you ready to open a PR from your branch?

**Comment by MSP-Greg** [comment]
@sodabrew Almost.  One last item.  I just decided to add some output similar to

```
no IPv6 in this host, skipping tests in C:/projects/eventmachine/tests/test_ipv6.rb
EM.popen not implemented, skipping tests in C:/projects/eventmachine/tests/test_processes.rb
EM Library Type: extension
```

with SSL information.  Is there any way to determine what version of OpenSSL EM was compiled with?  I looked over constants and things, couldn't see anything.

Over in Windows world, we've has some issues where Ruby OpenSSL gem was updated with a different version of OpenSSL than was originally used for the base dll's when Ruby was built...
